apiserver/common/crossmodel/bakery.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package crossmodel
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery/checkers"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/httpbakery"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+	"gopkg.in/macaroon.v2"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/apiserver/authentication"
+	"github.com/juju/juju/apiserver/bakeryutil"
+	coremacaroon "github.com/juju/juju/core/macaroon"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/state/bakerystorage"
+)
+
+// OfferBakery is a bakery service for offer access.
+type OfferBakery struct {
+	clock clock.Clock
+
+	bakery         authentication.ExpirableStorageBakery
+	accessEndpoint string
+}
+
+// OfferBakeryInterface is the interface that OfferBakery implements.
+type OfferBakeryInterface interface {
+	Clock() clock.Clock
+	SetClock(clock.Clock)
+	Bakery() authentication.ExpirableStorageBakery
+	RefreshDischargeURL(string) error
+
+	GetConsumeOfferCaveats(offerUUID, sourceModelUUID, username string) []checkers.Caveat
+	InferDeclaredFromMacaroon(macaroon.Slice, map[string]string) map[string]string
+	CreateDischargeMacaroon(
+		context.Context, string, map[string]string, map[string]string, bakery.Op, bakery.Version,
+	) (*bakery.Macaroon, error)
+}
+
+// Clock returns the clock used by the bakery service.
+func (o *OfferBakery) Clock() clock.Clock {
+	return o.clock
+}
+
+// SetClock sets the clock used by the bakery service.
+func (o *OfferBakery) SetClock(clock clock.Clock) {
+	o.clock = clock
+}
+
+// Bakery returns the ExpirableStorageBakery service.
+func (o *OfferBakery) Bakery() authentication.ExpirableStorageBakery {
+	return o.bakery
+}
+
+// RefreshDischargeURL updates the discharge URL for the bakery service.
+func (o *OfferBakery) RefreshDischargeURL(accessEndpoint string) error {
+	o.accessEndpoint = accessEndpoint
+	return nil
+}
+
+// NewOfferBakeryForTest is for testing.
+func NewOfferBakeryForTest(bakery authentication.ExpirableStorageBakery, clk clock.Clock) *OfferBakery {
+	return &OfferBakery{bakery: bakery, clock: clk}
+}
```

Comment:
This goes in package_test.go not prod code

Reply:
This is because we have some existing tests in other packages import it.
apiserver/facades/controller/crossmodelrelations/crossmodelrelations_test.go
apiserver/facades/controller/crossmodelsecrets/crossmodelsecrets_test.go
apiserver/facades/client/applicationoffers/access_test.go